### ISSUE TITLE
Prevention of automatic flooding of the worker queue by using fake requests

### ISSUE THREAD

**Issue by johnassel** [friendica/friendica#12729]
Lately my instance went very slow and upon investigation I noticed that my worker queue was full (999 open jobs).
Nearly all of them were related to queries to randomly generated subdomains of activitypub-troll.cf - like fhfaja9ahf9303.activitypub-troll.cf. 
Each of the subdomains had one contact associated which was filling the database, too.
In fact I had permanently 999 worker jobs and a lot of junk in the database.
My solution then was to block the domain *.activitypub-troll.cf through bin/console as the webpage for blocking instances was overloaded when checking the pattern. Then I had to block and purge the subdomains in chunks like a*.activitypub-troll.cf to prevent overloading. According to this https://github.com/mastodon/mastodon/issues/21977 they are using an infinite loop to generate these massive amounts of requests.

A detection and prevention for issues like this would be great since as far as I know there is currently no such feature.
This could be a rate limit for each domain+subdomain which is high enough that normal use is not impacted.

This should reduce possible future attacks with different domains.

**Comment by DanielSiepmann** [comment]
TYPO3 an open source PHP CMS added Symfonies Rate Limiter to limit attack surface on logins. Maybe that's a possible approach here as well? https://symfony.com/doc/current/rate_limiter.html

**Comment by MrPetovan** [comment]
What version are you running? In the hotfix release 2023.01 we added a provision for this specific attack pattern.

**Comment by johnassel** [comment]
I'm on 2023.03-dev

**Comment by annando** [comment]
I guess that the idea is to automatically detect such threats. So the system could detect on their own that there is some domain with more than X subdomains, which is suspicious. So i could raise a flag.

**Comment by johnassel** [comment]
Yes, something like that. I bet it will be just a matter of time until we see more of these attacks.

**Comment by annando** [comment]
But there also could be services that provide hosted servers under a given subdomain. So this has to be considered as well.

**Comment by tek-aevl** [comment]
The second method is to call millions of file request till the workers max out causing lag.

**Comment by ledlamp** [comment]
you need to add a depth limit to recursive requests or make it not automatically resolve mentions in pinned message. the software sees an infinite tree of references and tries to crawl it forever.

**Comment by shiribailem** [comment]
For this one specifically and similar attacks, one suggestion is to also track primary domains in detection efforts and the number of different subdomains for the same domain are in the list.

So if the server has scanned 5 different subdomains under `activitypub-troll.cf` in a row and they don't resolve then maybe stop adding it for a while? Maybe even track the origin of them being added and flag that?

**Comment by 4nanook** [comment]
2023.09-dev and six months later this is still a problem (today a heavy problem) for me as well.  Now I've got a script deleting them once a minute but they are added at a rate of about 200/second so the queue literally fills up after just five seconds.  Also, it would be VERY helpful if the IP address of requests were logged then we could simply handle this at the apache or nginx level.

**Comment by shiribailem** [comment]
@4nanook if you block it before a delete, it shouldn't add anymore after

But I do think it would make sense to maybe add a routine to the block process where it scans the list and removes any entries involving the blocked server?

**Comment by Quix0r** [comment]
> 2023.09-dev and six months later this is still a problem (today a heavy problem) for me as well. Now I've got a script deleting them once a minute but they are added at a rate of about 200/second so the queue literally fills up after just five seconds. Also, it would be VERY helpful if the IP address of requests were logged then we could simply handle this at the apache or nginx level.

My PR that addressed this issue had been altered by @MrPetovan . The check of being blocked in `src/Model/GServer.php` method `setFailureByUrl()` also need to check this. Every "instance" from these domains (`activitypub-troll.cf`, etc.) is a failing instance and will be added. ID 0806fa3f37159b233650b4596696a916ad0f311c from my repository includes this removal again.

If you like, you can cherry-pick it but you need to `git remote add` my repository first, then `git fetch --all`.
